### ISSUE TITLE
Proposal: Reorganize compatibility data into one YAML file

### ISSUE THREAD

**Issue by Thunderforge** [scummvm/scummvm-web#121]
This is a proposal to reorganize the [compatibility data](https://github.com/scummvm/scummvm-web/tree/master/data/compatibility) on the ScummVM website. Feedback is encouraged.

## Problem
At the time of this writing, there are 40 different files detailing the compatibility of various versions of ScummVM. While this does make it easy to see what was supported in a particular version, there are some major downsides:

- Dozens of files makes updating difficult and allows for inconsistencies
   - Changing the XML structure can require changing up to 40 different files
   - If new metadata tags are added without changing the structure, then old compatibility pages have empty columns (e.g. [1.6.0 compatibility](https://www.scummvm.org/compatibility/1.6.0/) has an empty "Data Files" column)
   - Game names are inconsistent between pages (e.g. [2.1.1](https://www.scummvm.org/compatibility/2.1.1/) says "Indiana Jones and the Fate of Atlantis", but [0.4.0](https://www.scummvm.org/compatibility/0.4.0/) says "Indiana Jones 4 and the Fate of Atlantis")
   - Tags are inconsistent between pages (e.g. some have `<percent>`, others have `<support_level>`)
- The current `<support_level>` tag is not working well
  - The meaning of each level has been ambiguous/contentious (#102) and undocumented (#104)
  - The `Untested`, `Broken`, and `Bugged` levels are largely unused, and developers have [indicated](https://github.com/scummvm/scummvm-web/pull/102#issuecomment-544666480) that they don't want to continue to use them (and PRs to add them like #107 have been rejected)
  - Having two levels of distinction (`Good` and `Excellent`) means it's more difficult to contrast different levels of support
- XML is not a preferred data format, with YAML being a preferred alternative
  - The XML format is not suitable for sorting or categorization (#110)
  - Data for other pages on the website have been converted to the YAML format (e.g. Game Demos via #116), and this is a trend we seem to want to continue

### A new `since` metadata value
A new `since` metadata value would be included with each game entry (matching the [value on the Wiki](https://wiki.scummvm.org/index.php?title=Maniac_Mansion)) that would indicate which version of ScummVM added support. For instance, Maniac Mansion would have `<since>0.5.0</since>` (YAML: `since: "0.5.0"`).

This would *replace* both the `<percentage>` tag and the `<support_level>` tag. While this does mean that data is lost (see "Drawbacks" section), it also means that there are no more debates on what to classify each version. Anybody who wants to know if there are lots of issues or not can simply look at the bug tracker.

I checked a few games that are listed as `Bugged` or `Broken`, and the Wiki says they are unsupported, and therefore won't have a value for `since`. I suggest they all be removed.

### One file for all games ever supported
The `since` value allows us to have *one* file for compatibility containing all games. It would contain all games that have been supported in any version of ScummVM. Then the list can be filtered (via PHP or in the frontend) to whichever ones are supported in the requested version.

For instance, if the user wants to see which ones are supported in ScummVM version 0.5.0, the full list would be filtered to any games that have a `since` value of 0.5.0 or earlier. The `CompatabilityModel.php` file already includes a `compareVersions()` plugin to compare semantic versions, so I don't think there is any additional work needed.

### YAML file format

The above could be done in XML, but this would be a good time to switch over to YAML. This would also give us the opportunity to add a few fields such as `engine` or `versionsSupported`. For instance:

```
- game:
    name: Maniac Mansion
    shortName: maniac
    company: LucasArts
    engine: SCUMM
    datafiles: Maniac_Mansion_.28Original_or_Enhanced.29
    since: 0.5.0
    versionsSupported:
        - Amiga
          # If necessary, we can add since information to each version
           - since: x.y.z
        - Apple II
        - Atari ST
        - Commodore 64
        - Macintosh
        - NES
        - PC
```

## Advantages

 - With one file, keeping game information consistent between releases is easy
 - New metadata like `engine` can be added easily since there is only one file
 - Game names will be consistent between versions
 - There are no more debates on what to set the subjective `percentage` or `support_level` to, since it's replaced by the objective `since` that matches the Wiki
 - The YAML file format will allow for new website features like sortable tables and filtering (#110)

## Disadvantages

 - Historical data for `percentage` or `support_level` is lost (e.g. the website no longer shows that version 1.6.0 had 90% compatibility with Maniac Mansion)
   - I'm guessing that this isn't a big loss, but I don't know for sure. Perhaps site analytics or a straw poll in IRC/Discord would be helpful?
 - A fair amount of work is needed to make this conversion (I'm willing to volunteer)

**Comment by sev-** [comment]
You are missing another, much bigger disadvantage: you will be losing the historical notes, as those also change from version to version.

If you would create a schema which would look something like the following (in XML):
```
<game>
   <name/>
   <target/>
   <datafiles/>
   <added_in>0.5.0</added_in>
   <details>
       <entry>
           <since>0.12.0</since>
           <notes>Foo bar, Amiga and Atari versions are supported</notes>
       </entry>
       <entry>
           <since>0.8.0</since>
           <support_level>blah++</support_level>
           <notes>Foo bar, and Amiga</notes>
       </entry>
       <entry>
           <since>0.5.0</since>
           <support_level>blah</support_level>
           <notes>Foo bar</notes>
       </entry>
     </details>
```

Entries override the previous ones, e.g 0.12.0 from the example above takes support level from 0.8.0, or copy/paste if it is difficult to implement.

**Comment by sev-** [comment]
Actually. After re-reading your original idea more carefully, you're covering what I just mentioned, the only thing is to add  optional '- notes' to version platform entry.

I am willing to help you with converting the data.

**Comment by Thunderforge** [comment]
To summarize a Discord conversation between @sev and I:

* We are in agreement that only games that are supported (a.k.a. compatible), as determined by the engine author (and subsequently reflected in the ScummVM release notes and Wiki) should be on the list. So games like *Moonbase Commander* should not be in the compatibility data despite being playable because the engine author has determined that it's not supported due to broken networking code.
 * If a game is supported for some platforms, but not others, sev would like all platforms to be listed, with accompanying notes listing the status of the platform. He gave the following pseudocode:
```
- Amiga
   - since: 0.12.0
       - notes: fully supported
   - since: 0.10.0
       - notes: missing sound
```
 * So far, I've suggested removing the "notes" of defects and just delineating between supported and unsupported platforms
```
name: "FooGame"
since: "0.2.0"
platforms:
  - supported:
      - name: "Windows"
        since: "0.2.0" # Could be omitted if same as game's since
      - name: "Macintosh"
        since: "0.4.0"
  - unsupported:
      - name: "Amiga"
      - name: "FM-Towns"
```

We had to end our discussion, so the final point hasn't been resolved yet. Feedback from anybody, including other users, is appreciated!

**Comment by Mataniko** [comment]
I'd like to make a distinction that there is a difference between games being and in the data, vs games being displayed on the website.

As we discussed on Discord, my biggest concern is not the data model - but how to manage a data set that is growing in both size and complexity.

We are in agreement that we'd like to use data files in the repository. As an intermediary we should consider building an actual database for the site context, using a google sheet, which can be easily converted back to YAML. 

The sheet will allow us to cross reference data and enforce some validation, plus visualizing the data model as it evolves without requiring adjustment to the code.

See https://docs.google.com/spreadsheets/d/1QzwFleEKXOsE59cYMOcQB7C2f0Np48uAQOCG8kicX_s/edit#gid=1416148232 as an example for compatibility data plus a small subset of relevant data.

I don't think we need to maintain compatibility pages for each ScummVM version as we do now, instead the compatibility page should reference the latest DEV build, and the version history of each game support can exist within the game specific compatibility page

**Comment by Thunderforge** [comment]
> I'd like to make a distinction that there is a difference between games being and in the data, vs games being displayed on the website.

In what cases would games be in the data, but not be in the website? My thought was that only supported games would be listed. The Wiki already has [unsupported games](https://wiki.scummvm.org/index.php?title=Category:Unsupported_Games) and I don't know that it's desirable to reproduce it.

> We are in agreement that we'd like to use data files in the repository.

To clarify for any observers: this means that the end result will be files on GitHub that anybody can edit with a pull request, as we have now. Although anything intermediate may be done privately (e.g. direct messages on Discord), the end result will be public.

Your spreadsheet is great! It was a bit unclear, but I think it's a combination of all existing compatibility spreadsheets?

> I don't think we need to maintain compatibility pages for each ScummVM version as we do now, instead the compatibility page should reference the latest DEV build, and the version history of each game support can exist within the game specific compatibility page

Agreed. The final UI is still up for discussion, but as stated above, I want one single YAML file with all compatibility data. A lot of redundancy will be able to be removed thanks to that.

**Comment by Thunderforge** [comment]
A few things I noticed about going through the "Notes" section:

* The vast majority say things like "No known issues, game is completable." This seems redundant (I would expect a supported game to be completable).
* Things get *really* out of date because people aren't updating it. I found half a dozen games that had something like "The code is brand new and hasn't been thoroughly tested yet". The worst I saw was *Dragon History*, added in 1.1.0, which was in April 2010. So for *ten years* it's had this status.
* Sometimes notes refer to issues that hardly seem worth mentioning. E.g. *Putt Putt Saves the Zoo* says "Minor graphical glitches when meeting Kenya in HE72 version". That refers to [this bug](https://bugs.scummvm.org/ticket/9619#no1), which is "Normal" severity. Why bother mentioning it?

My takeaway is: we should remove as many notes as possible from this going forward. They become redundant, out of date, and often not worth mentioning.
